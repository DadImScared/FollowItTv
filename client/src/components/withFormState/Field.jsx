
import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class Field extends Component {

  updateField = (e) => {
    const { updateForm, id, validators } = this.props;
    updateForm(e, id, validators);
  };

  render() {
    return React.cloneElement(this.props.children, { onChange: this.updateField });
  }
}

Field.propTypes = {
  id: PropTypes.string.isRequired,
  validators: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  updateForm: PropTypes.func.isRequired
};
