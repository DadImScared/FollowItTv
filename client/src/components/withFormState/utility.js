
export const removeValidateString = (data, key, errors) => {
  if (key in errors) {
    if (errors[key].includes(`, ${data[key]}`)) { // error string is in middle somewhere or at end
      errors[key] = errors[key].replace(`, ${data[key]}`, '');
    }
    else if (errors[key] === data[key]) { // error last remaining
      delete errors[key];
    }
    else { // error string is at start
      errors[key] = errors[key].replace(`${data[key]}, `, '');
    }
  }
};

export const applyValidation = async (props, state, id, validators) => {
  const initialErrors = { ...state.formErrors };
  const newErrors = {};

  for (const validator of validators) {
    const { remove, ...data } = await validator(props, state, id);
    Object.keys(data).forEach((key) => {
      if (remove) {
        removeValidateString(data, key, initialErrors);
      }
      else {
        if (key in initialErrors) {
          if (!initialErrors[key].includes(data[key])) {
            initialErrors[key] += `, ${data[key]}`;
          }
        }
        else {
          if (key in newErrors) {
            newErrors[key] += `, ${data[key]}`;
          }
          else {
            newErrors[key] = data[key];
          }
        }
      }
    });
  }
  return { ...initialErrors, ...newErrors };
};
