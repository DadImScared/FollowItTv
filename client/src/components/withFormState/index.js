
import React, { Component } from 'react';

import _ from 'lodash';

import { applyValidation } from './utility';

function getDisplayName(WrapperComponent) {
  return WrapperComponent.displayName || WrapperComponent.name || 'Component';
}

export default function(WrapperComponent) {
  class WithFormState extends Component {
    constructor(...args) {
      super(...args);
      this.state = {
        form: {},
        formErrors: {},
        afterSend: {}
      };
    }

    validateDebounce = _.debounce(async (props, { formErrors, form }, id, updateFormErrors, validators) => {
      const errors = await applyValidation(props, { formErrors, form }, id, validators);
      updateFormErrors(errors);
    }, 200);

    updateForm = ({ target: { value } }, id, validators) => {
      const { afterSend } = this.state;
      // clears errors on field if in after send
      // if afterSend[id] === undefined that means the form was
      // submitted with empty fields which include field under 'id' but it still has an error
      // that we need to clear from the server validation
      if (
        this.state.afterSend[id]
        || this.state.afterSend[id] === undefined
        || (typeof afterSend[id] === 'string' && !afterSend[id])
      ) {
        this.clearById(id);
      }
      const newForm = { ...this.state.form };
      newForm[id] = value;
      this.setState({ form: newForm }, () => {
        if (validators && validators.length) {
          this.validateDebounce(this.props, this.state, id, this.updateFormErrors, validators);
        }
      });
    };

    clearFields = () => {
      this.setState({ form: {} });
    };

    updateFormErrors = (newErrors) => this.setState({ formErrors: newErrors });

    handleErrorResponse = (data) => {
      const newErrors = {};
      Object.keys(data).forEach((key) => {
        if (key === 'non_field_errors') {
          newErrors['nonFieldErrors'] = data[key].join(', ');
        }
        else {
          if (key === 'detail') {
            newErrors['nonFieldErrors'] = data[key];
          }
          else {
            newErrors[key] = typeof data[key] === 'string' ? data[key]:data[key].join(', ');
          }
        }
      });
      this.updateFormErrors(newErrors);
    };

    clearErrors = () => {
      this.setState({ formErrors: {} });
    };

    clearById = (id) => {
      const afterSend = { ...this.state.afterSend };
      const formErrors = { ...this.state.formErrors };
      delete formErrors[id];
      afterSend[id] = false;
      this.setState({ afterSend, formErrors });
    };

    afterSend = () => {
      this.setState({ afterSend: Object.keys(this.state.form).length ? this.state.form:{} });
    };

    clear = () => {
      this.setState({ form: {}, formErrors: {} });
    };

    render() {
      const { afterSend, ...passedState } = this.state;
      return (
        <WrapperComponent
          updateForm={this.updateForm}
          updateFormErrors={this.updateFormErrors}
          handleErrorResponse={this.handleErrorResponse}
          clearErrors={this.clearErrors}
          clearFields={this.clearFields}
          clear={this.clear}
          submit={this.afterSend}
          {...passedState}
          {...this.props}
        />
      );
    }
  }
  WithFormState.displayName = `WithFormState(${getDisplayName(WrapperComponent)})`;
  return WithFormState;
}
