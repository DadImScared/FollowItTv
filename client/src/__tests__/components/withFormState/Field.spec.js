
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';

import TextField from '@material-ui/core/TextField';

import { Field } from '../../../components/withFormState/Field';

describe('Field', () => {
  let props;

  beforeEach(() => {
    props = {
      id: 'field 1',
      updateForm: sinon.spy()
    };
  });

  it('should render', () => {
    const wrapper = shallow(<Field {...props}><TextField value={'taco'}/></Field>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
