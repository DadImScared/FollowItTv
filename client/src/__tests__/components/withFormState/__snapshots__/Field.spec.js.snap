// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field should render 1`] = `
<TextField
  onChange={[Function]}
  required={false}
  select={false}
  value="taco"
/>
`;
